cupyx/scipy/sparse/index.py

```diff
@@ -0,0 +1,854 @@
+complex64_dtype = dtype('complex64')
+complex128_dtype = dtype('complex128')
+
+module_options = ('--std=c++11',)
+
+
+def _broadcast_arrays(a, b):
+    """
+    Same as cupy.broadcast_arrays(a, b) but old writeability rules.
+    NumPy >= 1.17.0 transitions broadcast_arrays to return
+    read-only arrays. Set writeability explicitly to avoid warnings.
+    Retain the old writeability rules, as our Cython code assumes
+    the old behavior.
+    """
+    x, y = cupy.broadcast_arrays(a, b)
+
+    # Writeable doesn't seem to exist on flags in cupy
+    # x.flags.writeable = a.flags.writeable
+    # y.flags.writeable = b.flags.writeable
+    return x, y
+
+
+bin_col_offsets_ker_types = {dtype('int32'): 'bin_col_offsets_ker_str<int>'}
+bin_col_offsets_ker = core.RawModule(code="""
+    template<typename I>
+    __global__ void bin_col_offsets_ker_str(I n_idx,
+                                            I *col_idxs,
+                                            I *col_offsets) {
+
+        // Get the index of the thread
+        I jj = blockIdx.x * blockDim.x + threadIdx.x;
+
+        if(jj < n_idx) {
+            const I j = col_idxs[jj];
+            atomicAdd(col_offsets+j, 1);
+        }
+    }
+""", options=module_options, name_expressions=tuple(
+    bin_col_offsets_ker_types.values()))
+
+
+def bin_col_offsets(n_idx, col_ids, col_offsets, tpb=32):
+    grid = math.ceil(n_idx / tpb)
+    kernel = bin_col_offsets_ker.get_function(
+        bin_col_offsets_ker_types[int32_dtype])
+    kernel((grid,), (tpb,), (n_idx, col_ids, col_offsets))
+
+
+csr_column_index1_ker_types = {int32_dtype: 'csr_column_index1_ker<int>'}
+csr_column_index1_ker = core.RawModule(code="""
+    template<typename I>
+    __global__ void csr_column_index1_ker(I n_row,
+                                   I *col_offsets,
+                                   I *Ap,
+                                   I *Aj,
+                                   I *Bp) {
+
+        // Get the index of the thread
+        I i = blockIdx.x * blockDim.x + threadIdx.x;
+
+        if(i < n_row) {
+
+            I new_col_size = 0;
+
+            for(I jj = Ap[i]; jj < Ap[i+1]; jj++)
+                new_col_size += col_offsets[Aj[jj]];
+
+            Bp[i+1] = new_col_size;
+        }
+}
+""", options=module_options, name_expressions=tuple(
+    csr_column_index1_ker_types.values()))
+
+
+def csr_column_index1_degree(n_row, col_offsets, Ap, Aj, Bp, tpb=32):
+    grid = math.ceil(n_row / tpb)
+    kernel = csr_column_index1_ker.get_function(
+        csr_column_index1_ker_types[int32_dtype]
+    )
+    kernel((grid,), (tpb,),
+           (n_row, col_offsets, Ap, Aj, Bp))
+
+
+def csr_column_index1(n_idx, col_idxs, n_row, n_col,
+                      indptr, indices, offsets, new_indptr):
+    bin_col_offsets(n_idx, col_idxs, offsets)
+    csr_column_index1_degree(n_row, offsets, indptr, indices, new_indptr)
+
+    cupy.cumsum(offsets, out=offsets)
+    cupy.cumsum(new_indptr, out=new_indptr)
```

Comment:
Could you write a comment and rearrange the arguments? I feel them a little complex to understand whether each parameter is updated or not. How about allocating the memory of `offsets` and `res_indptr` here as return values?
```suggestion
def csr_column_index1(n_idx, col_idxs, n_row, indptr, indices):
    offsets = cupy.zeros(n_col, indices.dtype)
    res_indptr = cupy.zeros_like(indptr)
    bin_col_offsets(n_idx, col_idxs, offsets)
    csr_column_index1_degree(n_row, offsets, indptr, indices, new_indptr)
    cupy.cumsum(offsets, out=offsets)
    cupy.cumsum(new_indptr, out=new_indptr)
    return offsets, new_indptr
```

Reply:
I updated all the functions in this file that are called by `compressed.py` to allocate an return, rather than accept empty arrays and modify them. Please let me know if you'd still like me to document these.

Reply:
Also `n_idx` can be removed because it is equal to `indices.size`.